mpi.h

```diff
@@ -196,258 +196,261 @@ typedef struct {
-    MPI_GRAPH      = 202,
-    MPI_DIST_GRAPH = 203,
-
-    // Datatype Decoding Combiners
-    MPI_COMBINER_NAMED              = 301,
-    MPI_COMBINER_DUP                = 302,
-    MPI_COMBINER_CONTIGUOUS         = 303,
-    MPI_COMBINER_VECTOR             = 304,
-    MPI_COMBINER_HVECTOR            = 305,
-    MPI_COMBINER_INDEXED            = 306,
-    MPI_COMBINER_HINDEXED           = 307,
-    MPI_COMBINER_INDEXED_BLOCK      = 308,
-    MPI_COMBINER_HINDEXED_BLOCK     = 309,
-    MPI_COMBINER_STRUCT             = 310,
-    MPI_COMBINER_SUBARRAY           = 311,
-    MPI_COMBINER_DARRAY             = 312,
-    MPI_COMBINER_F90_REAL           = 313,
-    MPI_COMBINER_F90_COMPLEX        = 314,
-    MPI_COMBINER_F90_INTEGER        = 315,
-    MPI_COMBINER_RESIZED            = 316,
-    MPI_COMBINER_VALUE_INDEX        = 317,
+    MPI_WIN_FLAVOR_CREATE              = 321,
+    MPI_WIN_FLAVOR_ALLOCATE            = 322,
+    MPI_WIN_FLAVOR_DYNAMIC             = 323,
+    MPI_WIN_FLAVOR_SHARED              = 324,
 
     // File Positioning
-    MPI_SEEK_CUR                = 601,
-    MPI_SEEK_END                = 602,
-    MPI_SEEK_SET                = 603,
-
-    // Fortran Datatype Matching
-    MPIX_TYPECLASS_LOGICAL      = 801,
-    MPI_TYPECLASS_INTEGER       = 802,
-    MPI_TYPECLASS_REAL          = 803,
-    MPI_TYPECLASS_COMPLEX       = 804,
+    MPI_SEEK_SET                       = 401,
+    MPI_SEEK_CUR                       = 402,
+    MPI_SEEK_END                       = 403
+};
 
-    // Communicator Split Types
-    MPI_COMM_TYPE_SHARED          = 1001,
-    MPI_COMM_TYPE_HW_UNGUIDED     = 1002,
-    MPI_COMM_TYPE_HW_GUIDED       = 1003,
-    MPI_COMM_TYPE_RESOURCE_GUIDED = 1004,
-
-    // Predefined Communicator Attribute Keys
-    MPI_TAG_UB          = 10001,
-    MPI_IO              = 10002,
-    MPI_HOST            = 10003, // deprecated since MPI-4.1
-    MPI_WTIME_IS_GLOBAL = 10004,
-    MPI_APPNUM          = 10005,
-    MPI_LASTUSEDCODE    = 10006,
-    MPI_UNIVERSE_SIZE   = 10007,
-
-    // Predefined Windows Attribute Keys
-    MPI_WIN_BASE            = 20001,
-    MPI_WIN_DISP_UNIT       = 20002,
-    MPI_WIN_SIZE            = 20003,
-    MPI_WIN_CREATE_FLAVOR   = 20004,
-    MPI_WIN_MODEL           = 20005
+// Predefined Attribute Keys
+enum {
+    // Invalid Attribute Key
+    MPI_KEYVAL_INVALID                 = 0,
+
+    // Communicator
+    MPI_TAG_UB                         = 1,
+    MPI_IO                             = 2,
+    MPI_HOST                           = 3, // deprecated: MPI-4.1
+    MPI_WTIME_IS_GLOBAL                = 4,
+    MPI_UNIVERSE_SIZE                  = 5,
+    MPI_APPNUM                         = 6,
+    MPI_LASTUSEDCODE                   = 7,
+
+    // Window
+    MPI_WIN_BASE                       = 11,
+    MPI_WIN_DISP_UNIT                  = 12,
+    MPI_WIN_SIZE                       = 13,
+    MPI_WIN_CREATE_FLAVOR              = 14,
+    MPI_WIN_MODEL                      = 15
```

Comment:
👍 Implementors need to generate integer values for user-defined keyvals. Having predefined values in a small range makes it easier for implementation to generate values outside a predefined range.